subsys/bluetooth/host/sdp.c

```diff
@@ -2304,87 +2304,102 @@ static ssize_t sdp_get_seq_len_item(uint8_t **data, size_t len)
 
 static int sdp_get_uuid_data(const struct bt_sdp_attr_item *attr,
 			     struct bt_sdp_uuid_desc *pd,
-			     uint16_t proto_profile)
+			     uint16_t proto_profile,
+			     uint16_t proto_profile_index)
 {
 	/* get start address of attribute value */
 	uint8_t *p = attr->val;
 	ssize_t slen;
 
 	BT_ASSERT(p);
 
-	/* Attribute value is a SEQ, get length of parent SEQ frame */
-	slen = sdp_get_seq_len_item(&p, attr->len);
-	if (slen < 0) {
-		return slen;
-	}
-
-	/* start reading stacked UUIDs in analyzed sequences tree */
-	while (p - attr->val < attr->len) {
-		size_t to_end, left = 0;
-
-		/* to_end tells how far to the end of input buffer */
-		to_end = attr->len - (p - attr->val);
-		/* how long is current UUID's item data associated to */
-		slen = sdp_get_seq_len_item(&p, to_end);
-		if (slen < 0) {
-			return slen;
-		}
+	while (proto_profile_index > 0) {
+		/* start reading stacked UUIDs in analyzed sequences tree */
+		while (p - attr->val < attr->len) {
+			size_t to_end, left = 0;
+			ssize_t pre_slen = -1;
+
+			/* loop all the SEQ */
+			while (p - attr->val < attr->len) {
+				/* to_end tells how far to the end of input buffer */
+				to_end = attr->len - (p - attr->val);
+				/* how long is current UUID's item data associated to */
+				slen = sdp_get_seq_len_item(&p, to_end);
+				if (slen < 0) {
+					break;
+				}
+				pre_slen = slen;
+			}
```

Comment:
This is extremely hard to read, not only are there 3 nested while loops, but two of them are working on the same loop invariants? 

`p` is updated in the inner loop; don't you easily risk `p` becoming incompatible with the middle loop? 

I will strongly suggest to split this (and perhaps some of the below) into one or more new static functions.

Reply:
Updated, I separated a static function.